src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,97 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double CAPACITY_INCRICE = 1.5;
+    private int size;
+    private T[] elementT;
+
+    public ArrayList() {
+        elementT = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public void resize() {
+        int capacity = (int) (elementT.length * CAPACITY_INCRICE);
+        T[] tempArray = (T[]) new Object[capacity];
+        System.arraycopy(elementT, 0, tempArray, 0, size);
+        elementT = tempArray;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element on position " + index);
+        }
+        if (size == elementT.length) {
+            resize();
+        }
+        System.arraycopy(elementT, index, elementT, index + 1,size - index);
+        elementT[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index > size - 1) {
+            throw new ArrayListIndexOutOfBoundsException("No such index " + index);
+        } else {
+            return elementT[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index > size - 1) {
+            throw new ArrayListIndexOutOfBoundsException("Can't set element on position " + index);
+        } else {
+            elementT[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("No such index " + index);
+        }
```

Comment:
[Extract to the additional method and reuse it everywhere to avoid duplicated code ](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html#dont-create-repeating-code)